src/hooks/useRequest.ts

```diff
@@ -0,0 +1,79 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+
+export interface UseRequestOptions<T> {
+  initialValue?: T;
+  onError?: (err: Error) => void;
+  onSuccess?: (res: T) => void;
+}
+
+export interface UseRequestResponse<T> {
+  loading: boolean;
+  error: Error | undefined;
+  data: T | undefined;
+}
+
+export type UseRequestReturn<T> = [() => Promise<void>, UseRequestResponse<T>];
+
+/**
+ * Provides an API to statefully interact with promises. This hook
+ * prevents state updates if promise has not resolved before component unmounts.
+ * It will return a tuple with a request() function, and an object representing the state
+ */
+export function useRequest<T>(
```

Comment:
This is somewhat a "quick and dirty" implementation, though it is stable. I wanted to implement a standard way of interacting with api's in our client with access to load state & errors sooner than later. Once we have more clarity around our backend, we can review if we want to use a 3rd party dependency such as [react-query](https://react-query.tanstack.com/)